File: internal/auth/jwt.go
package auth

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/lestrrat-go/jwx/v3/jwk"
	"github.com/lestrrat-go/jwx/v3/jws"
	"github.com/lestrrat-go/jwx/v3/jwt"
)

// jwksRefresh bounds how long a cached JWKS is used (PRD A8: <= 1 h).
const jwksRefresh = time.Hour

// KeySetProvider resolves the signing keys for token verification.
// Production uses a refreshing JWKS cache; tests supply a static set.
type KeySetProvider interface {
	KeySet(ctx context.Context) (jwk.Set, error)
}

// KeySetFunc adapts a function to KeySetProvider.
type KeySetFunc func(ctx context.Context) (jwk.Set, error)

// KeySet implements KeySetProvider.
func (f KeySetFunc) KeySet(ctx context.Context) (jwk.Set, error) { return f(ctx) }

// Error is a bearer-token verification failure with a structured reason for
// logs. Only a generic 401 reaches the caller.
type Error struct{ Reason string }

func (e *Error) Error() string { return "token verification failed: " + e.Reason }

// Verifier validates Cognito access tokens (PRD A8).
type Verifier struct {
	issuer  string
	allowed map[string]struct{}
	keys    KeySetProvider
}

// NewVerifier configures a Verifier for issuer and the allow-listed client ids.
func NewVerifier(issuer string, allowedClientIDs []string, keys KeySetProvider) *Verifier {
	return &Verifier{issuer: issuer, allowed: scopeSet(allowedClientIDs), keys: keys}
}

// NewJWKSProvider returns a KeySetProvider that fetches the JWKS at jwksURL
// on first use and re-fetches once the cached set is older than jwksRefresh.
// Fetching is lazy so a cold start never blocks on Cognito before the first
// token arrives, and a transient fetch failure surfaces as a 401 rather than
// a crashed function.
func NewJWKSProvider(jwksURL string) KeySetProvider {
	return &jwksCache{url: jwksURL, ttl: jwksRefresh, client: &http.Client{Timeout: 10 * time.Second}, now: time.Now}
}

type jwksCache struct {
	url    string
	ttl    time.Duration
	client *http.Client
	now    func() time.Time

	mu        sync.Mutex
	set       jwk.Set
	fetchedAt time.Time
}

// KeySet implements KeySetProvider.
func (c *jwksCache) KeySet(ctx context.Context) (jwk.Set, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.set != nil && c.now().Sub(c.fetchedAt) < c.ttl {
		return c.set, nil
	}
	set, err := jwk.Fetch(ctx, c.url, jwk.WithHTTPClient(c.client))
	if err != nil {
		if c.set != nil {
			return c.set, nil // serve the stale set rather than fail closed on a blip
		}
		return nil, fmt.Errorf("fetch jwks %s: %w", c.url, err)
	}
	c.set, c.fetchedAt = set, c.now()
	return set, nil
}

// Verify checks signature (RS256 against the key set), iss, exp,
// token_use == "access", and an allow-listed client_id, then returns the
// caller with its scopes parsed from the space-separated scope claim.
func (v *Verifier) Verify(ctx context.Context, token string) (Principal, error) {
	set, err := v.keys.KeySet(ctx)
	if err != nil {
		return Principal{}, &Error{Reason: "unresolvable_signing_key"}
	}
	tok, err := jwt.Parse([]byte(token),
		jwt.WithKeySet(set, jws.WithInferAlgorithmFromKey(true)),
		jwt.WithIssuer(v.issuer),
		jwt.WithRequiredClaim("sub"),
		jwt.WithValidate(true),
	)
	if err != nil {
		return Principal{}, &Error{Reason: classify(err)}
	}
	var tokenUse string
	if err := tok.Get("token_use", &tokenUse); err != nil || tokenUse != "access" {
		return Principal{}, &Error{Reason: "wrong_token_use"}
	}
	var clientID string
	if err := tok.Get("client_id", &clientID); err != nil {
		return Principal{}, &Error{Reason: "unknown_client"}
	}
	if _, ok := v.allowed[clientID]; !ok {
		return Principal{}, &Error{Reason: "unknown_client"}
	}
	var scope string
	_ = tok.Get("scope", &scope) // absent scope claim means no scopes
	sub, _ := tok.Subject()
	return Principal{
		Subject:  sub,
		ClientID: clientID,
		Scopes:   scopeSet(strings.Fields(scope)),
		Method:   "oauth",
	}, nil
}

func classify(err error) string {
	switch {
	case errors.Is(err, jwt.TokenExpiredError()):
		return "expired"
	case errors.Is(err, jwt.InvalidIssuerError()):
		return "wrong_issuer"
	case errors.Is(err, jwt.MissingRequiredClaimError()):
		return "missing_claim"
	default:
		return "invalid_token"
	}
}

File: src/aws_messaging_mcp/auth/jwt.py
"""Cognito access-token verification (PRD A8).

Tokens are validated for signature (RS256 against the pool's JWKS, cached at
most one hour), issuer, expiry, ``token_use == "access"``, and an allow-listed
``client_id``. Cognito access tokens carry no ``aud`` claim; the ``client_id``
claim plays that role.
"""

from __future__ import annotations

from collections.abc import Callable

import jwt as pyjwt

from aws_messaging_mcp.auth.principal import Principal

#: Resolves the signing key for a token. Production uses ``PyJWKClient``;
#: tests inject a resolver returning a PEM public key.
KeyResolver = Callable[[str], "pyjwt.PyJWK | str | bytes"]

_JWKS_CACHE_SECONDS = 3600


class AuthError(Exception):
    """A bearer token failed verification.

    Attributes:
        reason: Machine-readable failure reason for structured logs. Never
            echoed back to the caller beyond a generic 401.
    """

    def __init__(self, reason: str) -> None:
        """Store the failure reason.

        Args:
            reason: Machine-readable failure reason.
        """
        super().__init__(reason)
        self.reason = reason


class TokenVerifier:
    """Verifies Cognito access tokens and produces a ``Principal``."""

    def __init__(
        self,
        issuer: str,
        allowed_client_ids: tuple[str, ...],
        key_resolver: KeyResolver | None = None,
    ) -> None:
        """Configure the verifier.

        Args:
            issuer: Expected ``iss`` claim (the Cognito user-pool issuer URL).
            allowed_client_ids: App-client ids accepted in ``client_id``.
            key_resolver: Signing-key resolver; defaults to a caching
                ``PyJWKClient`` against the issuer's JWKS endpoint.
        """
        self._issuer = issuer
        self._allowed_client_ids = frozenset(allowed_client_ids)
        if key_resolver is None:
            jwks_client = pyjwt.PyJWKClient(
                f"{issuer}/.well-known/jwks.json",
                cache_keys=True,
                lifespan=_JWKS_CACHE_SECONDS,
            )
            key_resolver = jwks_client.get_signing_key_from_jwt
        self._resolve_key = key_resolver

    def verify(self, token: str) -> Principal:
        """Verify ``token`` and return the authenticated principal.

        Args:
            token: The raw bearer token.

        Returns:
            The verified caller.

        Raises:
            AuthError: On any verification failure, with a structured reason.
        """
        try:
            key = self._resolve_key(token)
        except Exception as exc:
            raise AuthError("unresolvable_signing_key") from exc
        try:
            claims = pyjwt.decode(
                token,
                key=key,
                algorithms=["RS256"],
                issuer=self._issuer,
                options={"require": ["exp", "iss", "sub"]},
            )
        except pyjwt.ExpiredSignatureError as exc:
            raise AuthError("expired") from exc
        except pyjwt.InvalidIssuerError as exc:
            raise AuthError("wrong_issuer") from exc
        except pyjwt.MissingRequiredClaimError as exc:
            raise AuthError("missing_claim") from exc
        except pyjwt.PyJWTError as exc:
            raise AuthError("invalid_token") from exc

        if claims.get("token_use") != "access":
            raise AuthError("wrong_token_use")
        client_id = claims.get("client_id")
        if not isinstance(client_id, str) or client_id not in self._allowed_client_ids:
            raise AuthError("unknown_client")

        scopes = frozenset(str(claims.get("scope", "")).split())
        return Principal(
            subject=str(claims["sub"]),
            client_id=client_id,
            scopes=scopes,
            method="oauth",
        )
